conversation/ShareJS
# Update socket.io to first non-vulnerable version

## Context

Existing version of `socket.io` and `socket.io-client` contains security vulnerabilities.
In particular, 2 critical ones related to `uglify-js`:

```
uglify-js  <=2.5.0
Severity: critical
Incorrect Handling of Non-Boolean Comparisons During Minification in uglify-js - https://github.com/advisories/GHSA-34r7-q49f-h37c
Regular Expression Denial of Service in uglify-js - https://github.com/advisories/GHSA-c9f4-xj24-8jqx
fix available via `npm audit fix --force`
Will install socket.io-client@4.4.0, which is a breaking change
node_modules/socket.io-client/node_modules/uglify-js
  socket.io-client  0.7.4 - 0.9.17
  Depends on vulnerable versions of uglify-js
  Depends on vulnerable versions of xmlhttprequest
  node_modules/socket.io-client
    socket.io  <=2.3.0
    Depends on vulnerable versions of socket.io-client
    node_modules/socket.io
```

Upgrade is required to fix critical security vulnerabilities related to `uglify-js`.

## Change

To take the least risky path possible, change introduces updates for `socket.io` and `socket.io-client` to **first** non-vulnerable versions (2.4.1 and 2.4.0 respectively, not to the latest ones, which are 4.4.0).

`npm audit --production` BEFORE change:
```
24 vulnerabilities (3 low, 3 moderate, 16 high, 2 critical)
```

`npm audit --production` AFTER change:
```
20 vulnerabilities (3 low, 3 moderate, 14 high)
```

## Note on bundle size

Version 0.8.7 (current): https://bundlephobia.com/package/socket.io-client@0.8.7 - 30kb compressed
Version 2.4.0 (proposed): https://bundlephobia.com/package/socket.io-client@2.4.0 - 18kb compressed

So we should expect some reduction in bundle size.

## How to test

Check if update to ShareJS works across TC and tc-share:

1. First, make sure that `./bin/exampleserver` works as expected
2. On local dev, update `sharejs` in TC with this ShareJS branch (`npm i share@github:conversation/ShareJS#32b6d7e`)
3. On local dev, update `tc-share` with this ShareJS branch
4. Follow these instructions https://github.com/conversation/tc-share#running-alongside-tc to run updated `tc-share` with TC
5. Make sure collaboration feature during editing is working as expected
6. Confirm that `npm audit --production` for TC no longer reports critical vulnerability coming from `uglify-js`
